man/man1/fgelev.1
.\" Copyright (C) 2017 Alessandro Menti
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
.\" Or try here: http://www.fsf.org/copyleft/gpl.html
.\"
.TH FGELEV 1 2020-11-15 FlightGear "FlightGear man pages"
.SH NAME
fgelev \- Compute FlightGear scenery elevation for a given list of points
.SH SYNOPSIS
\fBfgelev\fR [\fB\-\-expire\fR \fInum\fR] [\fB\-\-print\-solidness\fR]
[\fB\-\-fg\-root\fR \fIrootdir\fR] [\fB\-\-fg\-scenery\fR \fIscenerydir\fR]
[--tile-file osgbfilename] [--use-vpb]
.SH DESCRIPTION
.B fgelev
is a standalone utility that, given a list of points on standard input, prints
the corresponding elevation for each of them on standard output. It can also
(optionally) print whether the material covering that point is solid or not.

The list of points must be in the following form:

    \fIid lon lat\fR

where \fIid\fR is an arbitrary identifier for the point, \fIlon\fR is the
longitude of the point and \fIlat\fR is the latitude of the point. Positive
longitudes (latitudes) correspond to points located in the Eastern (Northern)
hemisphere; negative ones correspond to points located in the Western
(Southern) hemisphere.

The list of elevations returned by
.B fgelev
is in the following form:

    \fIid: elev solid\fR

where \fIid\fR is the identifier of the point passed as input, \fIelev\fR is
the elevation of the point in meters (or \fB-1000\fR if the elevation of the
given point was not found) and \fIsolid\fR is:
.IP \(bu 3
the string
.B solid
if the parameter
.B \-\-print\-solidness
was passed to
.B fgelev
and the material covering the point is solid;
.IP \(bu 3
the string
.B \-
if the parameter
.B \-\-print\-solidness
was passed to
.B fgelev
and the material covering the point is not solid;
.IP \(bu 3
absent if the parameter
.B \-\-print\-solidness
was not passed to \fBfgelev\fR.
.SH OPTIONS
.TP
\fB\-\-expire\fR \fInum\fR
To speed up elevation data retrieval,
.B fgelev
uses an internal cache. This option lets the user specify the number of
requests after which, if a point was not queried in them, it should be marked
as expired. By default,
.B fgelev
expires points not queried in the last \fB10\fR requests.
.TP
\fB\-\-print\-solidness\fR
Require
.B fgelev
to print the solidness of the material covering the queried point. See the
.B DESCRIPTION
section for more details.
.TP
\fB\-\-fg\-root\fR \fIrootdir\fR
Set the FlightGear data root directory (\fB$FG_ROOT\fR) to \fIrootdir\fR. If
this option is not set,
.B fgelev
uses the path specified in the
.B FG_ROOT
environment variable or, absent that, the FlightGear data directory chosen at
the time the program was compiled.
.TP
\fB\-\-fg\-scenery\fR \fIscenerydir\fR
Set the FlightGear scenery directory (\fB$FG_SCENERY\fR) to \fIscenerydir\fR.
If this option is not set,
.B fgelev
uses the path specified in the
.B FG_SCENERY
environment variable or, absent that, the
.B Scenery
subdirectory inside \fB$FG_ROOT\fR.
.TP
\fB\-\-tile\-file\fR \fIosgbfilename\fR
Use a specific tile file when searching for elevation details, instead
of scanning through the whole world.
.TP
\fB\-\-use\-vpb\fR
If specified, enables WS3.0 scenery behaviour, searching under the vpb/
subdirectory.
.SH "EXIT STATUS"
.B fgelev
exits with
.B EXIT_SUCCESS
on success, with
.B EXIT_FAILURE
if it is unable to read data from standard input or to load the scenery.
.SH ENVIRONMENT
.IP "\fBFG_ROOT\fR" 4
If
.B FG_ROOT
is set and
.B \-\-fg\-root
is not set, it specifies the root data directory to use.
.IP "\fBFG_SCENERY\fR" 4
If
.B FG_SCENERY
is set and
.B \-\-fg\-scenery
is not set, it specifies the scenery directory to use.
.SH "SEE ALSO"
.BR fgfs (1)


NEWS
Version 2.0.0 - February 25, 2010

  FlightGear 2.0.0. reflects the maturation of the OpenSceneGraph port
  that started with the previous 1.9.0 release. In addition to many
  internal code improvements, FlightGear 2.0.0. marks the introduction
  of many new exciting improvements in the graphics and sound system,
  as well as improved usability of key features, and improved behavior
  of exsisting features. Highlights of this new version include:
  Dramatic new 3D clouds, dramatic lighting conditions, improved
  support for custom scenery, and many many new and detailed aircraft
  models.

  Sound 
  * Complete overhaul of the sound code
  * doppler effects
  * distance attenuation
  * 3D positional sound sources
  * assignment of sound sources to external objects (i.e. AI controlled
    aircraft)
  * User selection of the sound device

  Visual Effects
  * Use of Shaders for dynamic textures
  * Use of Effects files
  * Improved 3D clouds
  * Color changes based on humidity and other weather effects allow for very
    dramatic lighting conditions
  * Dynamic water textures
  * Text animation based on OSGText

  Usability
  * Allow screenshots in more common file formats
  * User selectable sound device
  * More intuitive selection of the weather settings through the GUI and/or
    commandline

  Infrastructure
  * Airport geometry data can be read from the scenery, allowing for more
    flexible regeneration of terrain tiles

  Internals
  * Improved efficiency of the property tree
  * A more efficient ground cache
  * Many improvements to the route management code
  * Removed many compiler warnings
  * More realistic atmosphere model

  Behavior
  * More realistic ILS behavior
  * Autopilot improvements
  * A generic autobrake function
  * Winds over mountainous areas cause up- and downdrafts that can be used
    for gliding
  * More realistic behavior of the route manager
  * Wild fires, which can be extinguished by firefighter aircraft operating
    across the multplayer server
  * Navaid frequencies and radials can be transmitted to Atlas

  Utilities
  * A python script to visualize Yasim configuration files in Blender

  AI
  * Allow traffic departing and arriving at the same airport
  * Add Ground Vehicles - including automobiles, trucks, articulated trucks,
    trains (including high speed trains)
  * ATC interactions between AI aircraft and ground controllers
  * Performance characteristics of AI aircraft can be specified in a
    performance database
  * Push-back vehicles are available for a selected number of aircraft
  * Add escorts for AI carrier - frigates, guided missile cruiser, amphibious
    warfare ships now make up the Vinson Battle Group
  * Improved radar functionality - now detects AI escorts etc.
  * AI objects are now solid (i.e. users can collide with them)
  * Some preliminary support for SID/STAR procedures for AI aircraft

Version 1.9.0
* December 20, 2008 (source code snapshot release)


New in 0.9.10
* April 5, 2006

* Many improvements to the multi-player mode.  Other airplanes are now
  updated smoothly.  There is a really nifty google maps based moving
  map that anyone can use to track all online activity.
* AI planes have improved ground following and are smarter about
  selecting the active runway.
* Include the ability to send ATC and AI aircraft communication to a
  local festival server to be converted to spoken speech output.
* Added a detailed model of the Honeywell MK VIII EGPWS ground
  proximity warning system.
* Complete overhaul/redo of terrain textures based on USGS satellite imagery.
* Added background ATC chatter.
* Updated/improved rain cone rendering.
* Updated manual.

* Upgraded to JSBSim v2.0
* Fix a discrepancy in sign of twist and incidence values in YASim
  between code and documentation.
* Updates to YASim for more fine grained propeller modeling.
* UFO will now honor elevator trim value so that it will respond to
  the autopilot altitude hold module.
* UFO can adjust it's speed range so you can use it for super high
  speed world exploration, or extremly fine grained precise positioning.
* Much work and progress with carrier interaction and launch/recovery
  modeling.

* Add a feature that allows the end user to click any where on the
  terrain.  The system will compute a ground intersection and report
  back the lon, lat, and elevation of the click point.  The click
  point is reported in the property system so it can be read by just
  about anything including external programs.  Later we added the
  ability to place an object at the click point as well.
* Add a feature to play one-off wav files.  These are qeued and played
  sequentially.  This was then leveraged to build an ATC background
  chatter module in nasal that can be turned on/off at run time.  This
  could also be leveraged for tutorials or other interactive type
  scripts to play prerecorded messages at specific times or under
  specific conditions..
* Allow clock to report local time as well as universal.
* Add a model of a KLN89 GPS.
* Allow nav radios to be slaved to the gps.
* Allow enabling/disabling of menu entries.
* Replace old hardcoded frame rate display with a much more flexible
  nasal/gui version.
* Add a listener function to properties so that nasal code can be
  triggered whenever a property changes.  This is more efficient than
  polling.
* Add a save-on-exit option to save user preferences.
* Add support for point sprites.
* Add support for emissive layers (such as lights) on 2d panels.
* Updated gui styles.
* Updated airport and navaid data.
* Added a new Nasal/XML based tutorial system.

* Fix a bug in the metar (real weather) threading that could lead to
  long pauses and even crashes.
* Various compiler/platform fixes.
* Fix various valgrind complaints.
* Better use of exceptions in places.
* Fix the 2d compass texture handling.
* Many tweaks and improvements to our keyboard input handling system.

utils/fgelev/fgelev.cxx
// fgelev.cxx -- compute scenery elevation
//
// Copyright (C) 2009 - 2012  Mathias Froehlich
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <cstdlib>
#include <iomanip>

#include <osg/ArgumentParser>
#include <osg/Image>

#include <simgear/props/props.hxx>
#include <simgear/props/props_io.hxx>
#include <simgear/misc/sg_path.hxx>
#include <simgear/misc/ResourceManager.hxx>
#include <simgear/bvh/BVHNode.hxx>
#include <simgear/bvh/BVHLineSegmentVisitor.hxx>
#include <simgear/bvh/BVHPager.hxx>
#include <simgear/bvh/BVHPageNode.hxx>
#include <simgear/bvh/BVHMaterial.hxx>
#include <simgear/scene/material/matlib.hxx>
#include <simgear/scene/model/BVHPageNodeOSG.hxx>
#include <simgear/scene/model/ModelRegistry.hxx>
#include <simgear/scene/util/SGReaderWriterOptions.hxx>
#include <simgear/scene/util/OptionsReadFileCallback.hxx>
#include <simgear/scene/util/SGSceneFeatures.hxx>
#include <simgear/scene/tgdb/userdata.hxx>

namespace sg = simgear;

class Visitor : public sg::BVHLineSegmentVisitor {
public:
    Visitor(const SGLineSegmentd& lineSegment, sg::BVHPager& pager) :
        BVHLineSegmentVisitor(lineSegment, 0),
        _pager(pager)
    { }
    virtual ~Visitor()
    { }
    virtual void apply(sg::BVHPageNode& node)
    {
        // we have a non threaded pager so load just right here.
        _pager.use(node);
        BVHLineSegmentVisitor::apply(node);
    }
private:
    sg::BVHPager& _pager;
};

// Short circuit reading image files.
class ReadFileCallback : public sg::OptionsReadFileCallback {
public:
    virtual ~ReadFileCallback()
    { }

    virtual osgDB::ReaderWriter::ReadResult readImage(const std::string& name, const osgDB::Options*)
    { return new osg::Image; }
};

static bool
intersect(sg::BVHNode& node, sg::BVHPager& pager,
          const SGVec3d& start, SGVec3d& end, double offset, const simgear::BVHMaterial** material)
{
    SGVec3d perp = offset*perpendicular(start - end);
    Visitor visitor(SGLineSegmentd(start + perp, end + perp), pager);
    node.accept(visitor);
    if (visitor.empty())
        return false;
    end = visitor.getLineSegment().getEnd();
    * material = visitor.getMaterial();
    return true;
}

int
main(int argc, char** argv)
{
    /// Read arguments and environment variables.

    // use an ArgumentParser object to manage the program arguments.
    osg::ArgumentParser arguments(&argc, argv);

    unsigned expire;
    if (arguments.read("--expire", expire)) {
    } else expire = 10;

    bool printSolidness = arguments.read("--print-solidness");

    std::string fg_root;
    if (arguments.read("--fg-root", fg_root)) {
    } else if (const char *fg_root_env = std::getenv("FG_ROOT")) {
        fg_root = fg_root_env;
    } else {
        fg_root = PKGLIBDIR;
    }

    SGPath fg_scenery;
    std::string s;
    if (arguments.read("--fg-scenery", s)) {
        fg_scenery = SGPath::fromLocal8Bit(s.c_str());
    } else if (std::getenv("FG_SCENERY")) {
        fg_scenery = SGPath::fromEnv("FG_SCENERY");
    } else {
        SGPath path(fg_root);
        path.append("Scenery");
        fg_scenery = path;
    }

    SGSharedPtr<SGPropertyNode> props = new SGPropertyNode;
    try {
        SGPath preferencesFile = fg_root;
        preferencesFile.append("defaults.xml");
        readProperties(preferencesFile, props);
    } catch (...) {
        // In case of an error, at least make summer :)
        props->getNode("sim/startup/season", true)->setStringValue("summer");

        SG_LOG(SG_GENERAL, SG_ALERT, "Problems loading FlightGear preferences.\n"
               << "Probably FG_ROOT is not properly set.");
    }

    bool use_vpb = arguments.read("--use-vpb");
    props->setBoolValue("/scenery/use-vpb", use_vpb);
    SGSceneFeatures::instance()->setVPBActive(use_vpb);

    /// now set up the simgears required model stuff

    simgear::ResourceManager::instance()->addBasePath(fg_root, simgear::ResourceManager::PRIORITY_DEFAULT);
    // Just reference simgears reader writer stuff so that the globals get
    // pulled in by the linker ...
    simgear::ModelRegistry::instance();

    sgUserDataInit(props.get());
    SGMaterialLibPtr ml = new SGMaterialLib;
    SGPath mpath(fg_root);
    mpath.append("Materials/default/materials.xml");
    try {
        ml->load(fg_root, mpath.local8BitStr(), props);
    } catch (...) {
        SG_LOG(SG_GENERAL, SG_ALERT, "Problems loading FlightGear materials.\n"
               << "Probably FG_ROOT is not properly set.");
    }
    simgear::SGModelLib::init(fg_root, props);

    // Set up the reader/writer options
    osg::ref_ptr<simgear::SGReaderWriterOptions> options;
    if (osgDB::Options* ropt = osgDB::Registry::instance()->getOptions())
        options = new simgear::SGReaderWriterOptions(*ropt);
    else
        options = new simgear::SGReaderWriterOptions;
    osgDB::convertStringPathIntoFilePathList(fg_scenery.local8BitStr(),
                                             options->getDatabasePathList());
    options->setMaterialLib(ml);
    options->setPropertyNode(props);
    options->setReadFileCallback(new ReadFileCallback);
    options->setPluginStringData("SimGear::FG_ROOT", fg_root);
    // we do not need the builtin boundingvolumes
    options->setPluginStringData("SimGear::BOUNDINGVOLUMES", "OFF");
    // And we only want terrain, no objects on top.
    options->setPluginStringData("SimGear::FG_ONLY_TERRAIN", "ON");

    string_list scenerySuffixes = {"Terrain"}; // Just Terrain
    options->setSceneryPathSuffixes(scenerySuffixes);

    props->getNode("sim/rendering/random-objects", true)->setBoolValue(false);
    props->getNode("sim/rendering/random-vegetation", true)->setBoolValue(false);

    std::string bvhFile = "w180s90-360x180.spt";
    if (arguments.read("--tile-file", s))
        bvhFile = s;

    // Here, all arguments are processed
    arguments.reportRemainingOptionsAsUnrecognized();
    arguments.writeErrorMessages(std::cerr);

    // Get the whole world bvh tree
    SGSharedPtr<sg::BVHNode> node;
    node = sg::BVHPageNodeOSG::load(bvhFile, options);

    // if no model has been successfully loaded report failure.
    if (!node.valid()) {
        SG_LOG(SG_GENERAL, SG_ALERT, arguments.getApplicationName()
               << ": No data loaded");
        return EXIT_FAILURE;
    }

    // We assume that the above is a paged database.
    sg::BVHPager pager;

    while (std::cin.good()) {
        // Increment the paging relevant number
        pager.setUseStamp(1 + pager.getUseStamp());
        // and expire everything not accessed for the past 30 requests
        pager.update(expire);

        std::string id;
        std::cin >> id;
        double lon, lat;
        std::cin >> lon >> lat;
        if (std::cin.fail())
            return EXIT_FAILURE;
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');

        SGVec3d start = SGVec3d::fromGeod(SGGeod::fromDegM(lon, lat, 10000));
        SGVec3d end = SGVec3d::fromGeod(SGGeod::fromDegM(lon, lat, -1000));

        const simgear::BVHMaterial* material = NULL;
        // Try to find an intersection
        bool found = intersect(*node, pager, start, end, 0, &material);
        double scale = 1e-5;
        while (!found && scale <= 1) {
            found = intersect(*node, pager, start, end, scale, &material);
            scale *= 2;
        }
        if (1e-5 < scale)
            std::cerr << "Found hole of minimum diameter "
                      << scale << "m at lon = " << lon
                      << "deg lat = " << lat << "deg" << std::endl;

        std::cout << id << ": ";
        string solid = material && material->get_solid() ? "solid" : "-";
        if (!found) {
            std::cout << "-1000" << std::endl;
        } else {
            SGGeod geod = SGGeod::fromCart(end);
            std::cout << std::fixed << std::setprecision(3) << geod.getElevationM();
            if( printSolidness )
                std::cout <<  " " << (material && material->get_solid() ? "solid" : "-");
            std::cout << std::endl;
        }
    }

    return EXIT_SUCCESS;
}
